runtime/src/tokio/blob_linux.rs

```diff
@@ -0,0 +1,245 @@
+            .join(hex(name));
+        let parent = match path.parent() {
+            Some(parent) => parent,
+            None => return Err(Error::PartitionCreationFailed(partition.into())),
+        };
+
+        // Create the partition directory if it does not exist
+        fs::create_dir_all(parent).map_err(|_| Error::PartitionCreationFailed(partition.into()))?;
+
+        // Open the file in read-write mode, create if it does not exist
+        let file = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(false)
+            .open(&path)
+            .map_err(|_| Error::BlobOpenFailed(partition.into(), hex(name)))?;
+
+        // Get the file length
+        let len = file.metadata().map_err(|_| Error::ReadFailed)?.len();
+
+        // Construct the blob
+        Ok(Blob::new(
+            partition.into(),
+            name,
+            file,
+            len as u32, // TODO danlaine: handle overflow
+        ))
+    }
+
+    async fn remove(&self, partition: &str, name: Option<&[u8]>) -> Result<(), Error> {
+        // Acquire the filesystem lock
+        let _guard = self.executor.fs.lock().await;
+
+        // Remove all related files
+        let path = self.executor.cfg.storage_directory.join(partition);
+        if let Some(name) = name {
+            let blob_path = path.join(hex(name));
+            fs::remove_file(blob_path)
+                .map_err(|_| Error::BlobMissing(partition.into(), hex(name)))?;
+        } else {
+            fs::remove_dir_all(path).map_err(|_| Error::PartitionMissing(partition.into()))?;
+        }
+        Ok(())
+    }
+
+    async fn scan(&self, partition: &str) -> Result<Vec<Vec<u8>>, Error> {
+        // Acquire the filesystem lock
+        let _guard = self.executor.fs.lock().await;
+
+        // Scan the partition directory
+        let path = self.executor.cfg.storage_directory.join(partition);
+        let mut entries = tokio::fs::read_dir(path)
+            .await
+            .map_err(|_| Error::PartitionMissing(partition.into()))?;
+        let mut blobs = Vec::new();
+        while let Some(entry) = entries.next_entry().await.map_err(|_| Error::ReadFailed)? {
+            let file_type = entry.file_type().await.map_err(|_| Error::ReadFailed)?;
+            if !file_type.is_file() {
+                return Err(Error::PartitionCorrupt(partition.into()));
+            }
+            if let Some(name) = entry.file_name().to_str() {
+                let name = from_hex(name).ok_or(Error::PartitionCorrupt(partition.into()))?;
+                blobs.push(name);
+            }
+        }
+        Ok(blobs)
+    }
+}
+
+impl crate::Blob for Blob {
+    async fn len(&self) -> Result<u64, Error> {
+        Ok(self.file.lock().await.2 as u64)
+    }
+
+    async fn read_at(&self, buf: &mut [u8], offset: u64) -> Result<(), Error> {
+        // Lock the file to ensure safe access
+        let mut foo = self.file.lock().await;
```

Comment:
Is it possible to have a read-write lock allowing concurrent readers?

Reply:
Once we migrate ring handling to a separate task, we will no longer need the lock for read or write.